File: tso_kernel/neurons.py
"""
LIF neurons and clusters — the physical substrate of TSO.

Each cluster corresponds to one conceptual node. Neurons within a cluster
share an embedding vector and fire together to signal concept activation.
"""
import numpy as np


class LIFNeuron:
    """Single leaky integrate-and-fire neuron."""
    def __init__(self):
        self.tau_m = 10.0
        self.v_rest = -65.0
        self.v_th = -55.0
        self.v_reset = -70.0
        self.v = self.v_rest
        self.refractory = 0

    def step(self, I_syn, dt=0.5):
        if self.refractory > 0:
            self.refractory -= 1
            self.spiked = False
            return 0.0
        dv = (dt / self.tau_m) * (-(self.v - self.v_rest) + I_syn)
        self.v += dv
        if self.v >= self.v_th:
            self.v = self.v_reset
            self.refractory = 2
            self.spiked = True
            return 1.0  # spike
        self.spiked = False
        return 0.0

    def reset(self):
        self.v = self.v_rest
        self.refractory = 0
        self.spiked = False


class LIFCluster:
    """A cluster of LIF neurons acting as one conceptual unit."""
    def __init__(self, n_neurons, label="", tau_m=10.0):
        self.n = n_neurons
        self.label = label
        self.tau_m = tau_m
        self.v_rest = -65.0
        self.v_th = -55.0
        self.v_reset = -70.0
        self.v = np.full(n_neurons, self.v_rest, dtype=np.float32)
        self.spikes = np.zeros(n_neurons, dtype=np.float32)
        self.rate = 0.0
        self.tau_rate = 50.0

    def step(self, I_syn, dt=0.5):
        dv = (dt / self.tau_m) * (-(self.v - self.v_rest) + I_syn)
        self.v += dv
        self.spikes.fill(0.0)
        fired = self.v >= self.v_th
        self.spikes[fired] = 1.0
        self.v[fired] = self.v_reset

        inst_rate = float(np.mean(self.spikes))
        a = 1.0 - np.exp(-dt / self.tau_rate)
        self.rate += a * (inst_rate - self.rate)
        return inst_rate

    def reset(self):
        self.v.fill(self.v_rest)
        self.spikes.fill(0.0)
        self.rate = 0.0

File: tso-kernel/src/neurons.rs
use ndarray::Array1;

#[derive(Clone)]
pub struct LIFNeuron {
    pub tau_m: f64,
    pub v_rest: f64,
    pub v_th: f64,
    pub v_reset: f64,
    pub v: f64,
    pub refractory: i32,
    pub spiked: bool,
}

impl LIFNeuron {
    pub fn new() -> Self {
        Self {
            tau_m: 10.0,
            v_rest: -65.0,
            v_th: -55.0,
            v_reset: -70.0,
            v: -65.0,
            refractory: 0,
            spiked: false,
        }
    }

    pub fn step(&mut self, i_syn: f64, dt: f64) -> f64 {
        if self.refractory > 0 {
            self.refractory -= 1;
            self.spiked = false;
            return 0.0;
        }
        let dv = (dt / self.tau_m) * (-(self.v - self.v_rest) + i_syn);
        self.v += dv;
        if self.v >= self.v_th {
            self.v = self.v_reset;
            self.refractory = 2;
            self.spiked = true;
            return 1.0;
        }
        self.spiked = false;
        0.0
    }

    pub fn reset(&mut self) {
        self.v = self.v_rest;
        self.refractory = 0;
        self.spiked = false;
    }
}

impl Default for LIFNeuron {
    fn default() -> Self {
        Self::new()
    }
}

#[derive(Clone)]
pub struct LIFCluster {
    pub n: usize,
    pub label: String,
    pub tau_m: f64,
    pub v_rest: f64,
    pub v_th: f64,
    pub v_reset: f64,
    pub v: Array1<f64>,
    pub spikes: Array1<f64>,
    pub rate: f64,
    pub tau_rate: f64,
}

impl LIFCluster {
    pub fn new(n_neurons: usize) -> Self {
        Self {
            n: n_neurons,
            label: String::new(),
            tau_m: 10.0,
            v_rest: -65.0,
            v_th: -55.0,
            v_reset: -70.0,
            v: Array1::from_elem(n_neurons, -65.0),
            spikes: Array1::zeros(n_neurons),
            rate: 0.0,
            tau_rate: 50.0,
        }
    }

    pub fn with_label(n_neurons: usize, label: &str) -> Self {
        let mut c = Self::new(n_neurons);
        c.label = label.to_string();
        c
    }

    pub fn step(&mut self, i_syn: &Array1<f64>, dt: f64) -> f64 {
        let dv = (dt / self.tau_m) * (-(&self.v - self.v_rest) + i_syn);
        self.v += &dv;
        self.spikes.fill(0.0);
        for i in 0..self.n {
            if self.v[i] >= self.v_th {
                self.spikes[i] = 1.0;
                self.v[i] = self.v_reset;
            }
        }
        let inst_rate = self.spikes.mean().unwrap_or(0.0);
        let a = 1.0 - (-dt / self.tau_rate).exp();
        self.rate += a * (inst_rate - self.rate);
        inst_rate
    }

    pub fn reset(&mut self) {
        self.v.fill(self.v_rest);
        self.spikes.fill(0.0);
        self.rate = 0.0;
    }
}
